src/merkle_tree/mod.rs

```diff
@@ -1,253 +1,399 @@
+    pub(crate) leaf_index: usize,
 }
 
-pub type Parameters<P> = <<P as Config>::H as FixedLengthCRH>::Parameters;
-pub type Digest<P> = <<P as Config>::H as FixedLengthCRH>::Output;
-
-impl<P: Config> Default for Path<P> {
-    fn default() -> Self {
-        let mut path = Vec::with_capacity(P::HEIGHT as usize);
-        for _i in 1..P::HEIGHT as usize {
-            path.push((
-                <P::H as FixedLengthCRH>::Output::default(),
-                <P::H as FixedLengthCRH>::Output::default(),
-            ));
-        }
-        Self { path }
+impl<P: Config> Path<P> {
+    /// The position of on_path node in `leaf_and_sibling_hash` and `non_leaf_and_sibling_hash_path`.
+    /// `position[i]` is 0 (false) iff `i`th on-path node from top to bottom is on the left.
+    ///
+    /// This function simply converts `self.leaf_index` to boolean array in big endian form.
+    fn position_list(&self) -> Vec<bool> {
+        let position: Vec<_> = (0..self.auth_path.len() + 1)
+            .map(|i| ((self.leaf_index >> i) & 1) != 0)
+            .rev()
+            .collect();
+        position
     }
 }
 
 impl<P: Config> Path<P> {
+    /// Verify that a leaf is at `self.index` of the merkle tree.
+    /// * `leaf_size`: leaf size in number of bytes
+    ///
+    /// `verify` infers the tree height by setting `tree_height = self.auth_path.len() + 2`
     pub fn verify<L: ToBytes>(
         &self,
-        parameters: &<P::H as FixedLengthCRH>::Parameters,
-        root_hash: &<P::H as FixedLengthCRH>::Output,
+        leaf_hash_parameters: &LeafParam<P>,
+        two_to_one_hash_parameters: &TwoToOneParam<P>,
+        root_hash: &TwoToOneDigest<P>,
         leaf: &L,
     ) -> Result<bool, crate::Error> {
-        if self.path.len() != (P::HEIGHT - 1) as usize {
-            return Ok(false);
-        }
-        // Check that the given leaf matches the leaf in the membership proof.
-        let mut buffer = [0u8; 128];
-
-        if !self.path.is_empty() {
-            let claimed_leaf_hash = hash_leaf::<P::H, L>(parameters, leaf, &mut buffer)?;
-
-            // Check if leaf is one of the bottom-most siblings.
-            if claimed_leaf_hash != self.path[0].0 && claimed_leaf_hash != self.path[0].1 {
-                return Ok(false);
-            };
+        // calculate leaf hash
+        let claimed_leaf_hash =
+            P::LeafHash::evaluate(&leaf_hash_parameters, &ark_ff::to_bytes!(&leaf)?)?;
+        // check hash along the path from bottom to root
+        let mut left_bytes;
+        let mut right_bytes;
+        if self.leaf_index & 1 == 0 {
+            // leaf is on left
+            left_bytes = ark_ff::to_bytes!(&claimed_leaf_hash)?;
+            right_bytes = ark_ff::to_bytes!(&self.leaf_sibling_hash)?;
+        } else {
+            // leaf is on right
+            left_bytes = ark_ff::to_bytes!(&self.leaf_sibling_hash)?;
+            right_bytes = ark_ff::to_bytes!(&claimed_leaf_hash)?;
+        };
```

Comment:
Can we extract this into a function called `select_left_right_bytes(index, input 1, input 2)`, and then re-use this in the later loop?

Reply:
That's awesome! I will do it now

Reply:
done